### DISCUSSION TITLE
[Enhancement] Align the XAML and Object-Model with WinUI as much as Possible

### DISCUSSION THREAD

**Discussion by robloo** [dotnet/maui#161] [General]
## Summary

This is a catch-all issue tracking specific changes needed to more closely align the Maui XAML and object model to WinUI. This originates from the discussion in #43 and #122.

The intention is for this to be a general discussion and to link (within the API Changes section) to specific proposals. The end goal, however difficult, should be to align Maui with WinUI so where features overlap the API is the same.

A starting list of differences was provided by @LanceMcCarthy at https://github.com/LanceMcCarthy/MauiSuggestions

## API Changes

The following issues track specific change proposals:

**Maui Repo**

 - [ ] #209 [Enhancement] Features for map
 - [x] #207 ~~Xaml Standardization[Enhancement]~~ (Closed, duplicate)
 - [ ] #188 [Enhancement] Fluent UI support 
 - [ ] #185 **[Enhancement] Make all controls fully customizable like in WPF or UWP**
 - [ ] #182 [Enhancement] Rating Control 
 - [ ] #177 [Enhancement] MAUI Font Icon 
 - [ ] #176 [Enhancement] Support DependencyObject.SetCurrentValue 
 - [ ] #165 [Enhancement] Support dual screen devices out of the box 
 - [ ] #158 [Enhancement] UWP as a first-class platform (alongside Android and iOS) 
 - [ ] #156 [Enhancement] Widgets / LiveTiles
 - [ ] #153 [Enhancement] What about Blend for Visual Studio Support MAUI 
 - [ ] #146 [Enhancement] Please get essential things first 
 - [ ] #138 [Enhancement] Rethink the layout system - Use "Items Repeater" everywhere 
 - [ ] #136 [Enhancement] Vector icons (SVG/XAML) 
 - [ ] #112 Question : What's the story for System.Drawing?
 - [ ] #110 [Enhancement] Replace some APIs with existing .NET ones
 - [ ] #99 DataContext in Xamarin And WPF
 - [ ] #98 [Enhancement] Better Desktop Experience
 - [ ] #97 [Enhancement] x:DataType should provide intellisense features
 - [ ] #85 [Enhancement] Replace Xamarin.Forms.Xaml with System.Xaml
 - [ ] #81 [Enhancement] Animation XAML Api
 - [ ] #78 [Enhancement] Decouple windowing and basic eventing?
 - [ ] #76 [Enhancement] The most wanted things!!!
 - [ ] #39 [Enhancement] Make all default spacings / margins zero
 - [ ] #12 [Spec] Border
 - [ ] #10 [Spec] Rounded Corners

**Xamarin.Forms Repo**

 - [x] :thumbsup: [#9178](https://github.com/xamarin/Xamarin.Forms/issues/9178) ~~[Spec] Shapes & Paths~~ (Added in XF 4.7)
 - [ ] [#7293](https://github.com/xamarin/Xamarin.Forms/issues/7293) [Spec] Brushes
 - [ ] [#2452](https://github.com/xamarin/Xamarin.Forms/issues/2452) Xamarin.Forms Drawing Spec
 - [ ] [#1856](https://github.com/xamarin/Xamarin.Forms/issues/1856) [Enhancement] Add brush support for background 

## Intended Use Case

Be a one-stop shop to discuss and track all issues related to aligning the Maui XAML and object model with WinUI.

**Comment by tomasfabian** [comment]
thank you @robloo

**Comment by charlesroddie** [comment]
By "XAML Object-Model" in the title I think you mean "XAML and Object-Model". I.e. this concerns MAUI .Net objects, not just the structure of XAML documents.

The current naming suggestions in https://github.com/LanceMcCarthy/MauiSuggestions don't seem good:
- Replacing Content with User in ContentView is not good. (A "User" control denoting a Control a "user" can inherit from, so "User" means non-sealed??) It's not necessary for alignment with WPF/UWP as there is a ContentControl there.
- VisualElement is better than FrameworkElement as Visual has a clear and appropriate meaning unlike Framework.
- For the "Controls", almost everything is just longer, so again this not an improvement.
- Only in Properties are there some improvements, with Height being better than HeightRequest, and VerticalAlignment being clearer than VerticalOptions.

The suggestions there are just copying of UWP/WPF without concern for appropriateness of naming.

**Comment by robloo** [comment]
@charlesroddie Everyone has an opinion and Microsoft, in my opinion, should not be changing XAML and object-model from existing .net technologies including WPF and UWP. There should be a XAML design board that aligns all the various dialects. Reinventing the wheel repeatedly is a primary issue with the Windows UI stacks. Consideration is not simply in how long the name of a property or control is. That said I'm getting off track, this was all discussed in #43 and doesn't belong here.

The comparison @LanceMcCarthy put together is a raw list of differences between UWP/XF. Just because a difference is called out in that list doesn't mean there is a request to change it in XF/Maui. The specific requests for changes are in the linked issues. This issue #161 simply collects all the specific proposals for changes that align with UWP.

**Comment by PapLabros** [comment]
At the 57:55 minute mark of the newest MAUI presentation: https://www.youtube.com/watch?v=7Gvr0d1vTQ4&ab_channel=dotNET, @davidortinau mentions that based on survey responses that they had, there wasn't enough pressure for them to align the XF XAML to the WPF/UWP one. That's pretty interesting, because based on every github discussion that I have read it seems that it's the other way around. Most of the discussions state that the XAML between frameworks should be consistent in order for the developers to be able to switch frameworks without having to learn them entirely from scratch and also to be able to port their existing applications easily.

Imho this is a terrible start for this new and promising framework not only because the port (re-write) existing WPF and UWP applications to MAUI will be way more difficult than it should be, but also because the WPF/UWP XAML feels way more consistent and better than the one XF currently have. Furthermore, since most of the big proposed changes to the XAML dialect are simple renames (and shouldn't really affect the existing users of XF, since they will simply apply the renames), I can't seem to find a reason to not do it. Unless there is a really important reason that discourages them from making the changes that we do not know of, this seems like a very poor decision. And no, the fact that with the new changes, Blend will not work with MAUI, is not a valid reason to not make them 😃 . That's just my opinion though!

**Reply by robloo** [reply]
What survey are they referring to? I certainly didn't hear of one to give my input. It may not have been a widely circulated survey which could have been on purpose.

The current theory from a discussion months ago here on GitHub is that Microsoft is only talking to their existing customers for feedback -- or heavily weighting feedback in their favor. They would rather cater to current XF users that attract new developers from existing XAML tooling. I also think there is a lot of politics in play at this point as well.

Personally, I would always want to hear from those that don't use my product as they will tell you what needs to be done to make it a more attractive for users. Microsoft isn't operating like that here and I really can't explain it other than they refuse to make breaking changes. Aside from Panos, I haven't seen any Microsoft leadership in this space that has vision. There is no one that can feel the product and instantly know what is a good/not-good idea to steer it in the right direction. It's a lot of design by committee and no one is willing to make changes if there aren't copious amounts of surveys and feedback first -- and surveys are unreliable and really only are useful to find gross inadequacies or gather new ideas. Anyway, it very much is a management problem at this point and Microsoft is killing their own ecosystem by fragmentation.

**Reply by PapLabros** [reply]
What's interesting is that the teams behind the ASP.NET Core and the C# language have done an excellent job. It's not that the Xamarin team has done a bad job, I still believe that MAUI is the right way to go. It's just that it can be a lot better, and a great cross platform UI framework supported by MS is the only thing that C# needs at this point. But it seems that having both an excellent back-end and a front-end framework with a great language to go along, would be too good to be true! 😛

**Reply by saint4eva** [reply]
Unifying the XAML dialects without creating a new one is not possible.

**Reply by charlesroddie** [reply]
> @robloo Microsoft is only talking to their existing customers for feedback -- or heavily weighting feedback in their favor. They would rather cater to current XF users that attract new developers from existing XAML tooling

"Developers from existing XAML tooling" is just a different circle of existing Microsoft customer. Naturally MS should weight XF users higher than WPF/UWP users when deciding on the future of XF. And a more important group is non-Microsoft customers, and they have never used XAML.

If you want to cater to existing WPF and UWP codebases you should write some conversion utility. It should not affect the object model itself unless those changes are intrinsically beneficial for XF.

**Reply by robloo** [reply]
This was all discussed before. Bottom line the original error was Xamarin deviating from existing dialects for XF. It was more tolerable until Maui was announced and now .net itself is going to have two different dialects. If XF was staying separate from .net I would be more inclined to ignore differences as I have in the past. I certainly agree with you though both sides have the same argument which is equally valid: don't change from convention. The issue was WPF already established convention for a lot of things (Width/Height, etc) and now its being changed for those of us using Xaml from the beginning (a greater number of developers). There really should be a unifying Xaml standard committee but Microsoft is happy to create fragmentation and inneficiencies for developers using more than one version of Xaml.

**Comment by Majestic-Mustang** [comment]
What's the update on this?

**Reply by robloo** [reply]
I'm not tracking this anymore and the general consensus in the linked issues was Microsoft generally was not going to align the XAML dialects. Xamarin Forms and Maui will go their own path. Additionally, we have the Avalonia dialect prominently now as well as WinUI not really taking off. So basically, it looks like these are never going to be aligned until one UI framework just clearly becomes dominant.

**Reply by Majestic-Mustang** [reply]
Yeah, that sucks. I was really hoping to use WPF XAML dialect in MAUI.

**Reply by ccarlo88** [reply]
The general consensus is don't use MAUI at all 😆
